src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -62,54 +62,74 @@ public function validate(mixed $entity, Constraint $constraint)
             throw new ConstraintDefinitionException('At least one field has to be specified.');
         }
 
-        if (null === $entity) {
+        if (null === $value) {
             return;
         }
 
-        if (!\is_object($entity)) {
-            throw new UnexpectedValueException($entity, 'object');
+        if (!\is_object($value)) {
+            throw new UnexpectedValueException($value, 'object');
         }
 
+        $entityClass = $value::class;
+
         if ($constraint->em) {
             $em = $this->registry->getManager($constraint->em);
 
             if (!$em) {
                 throw new ConstraintDefinitionException(sprintf('Object manager "%s" does not exist.', $constraint->em));
             }
         } else {
-            $em = $this->registry->getManagerForClass($entity::class);
+            $em = $this->registry->getManagerForClass($constraint->entityClass ?? $entityClass);
 
             if (!$em) {
-                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($entity)));
+                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($value)));
             }
         }
 
-        $class = $em->getClassMetadata($entity::class);
+        try {
+            $repository = $em->getRepository($entityClass);
+            $isEntity = true;
+        } catch (ORMMappingException|PersistenceMappingException) {
+            $isEntity = false;
+        }
 
-        $criteria = [];
-        $hasIgnorableNullValue = false;
+        if (null !== $constraint->entityClass) {
+            /* Retrieve repository from given entity name.
+             * We ensure the retrieved repository can handle the entity
+             * by checking the entity is the same, or subclass of the supported entity.
+             */
+            $repository = $em->getRepository($constraint->entityClass);
+            $supportedClass = $repository->getClassName();
 
-        foreach ($fields as $fieldName) {
-            if (!$class->hasField($fieldName) && !$class->hasAssociation($fieldName)) {
-                throw new ConstraintDefinitionException(sprintf('The field "%s" is not mapped by Doctrine, so it cannot be validated for uniqueness.', $fieldName));
+            if ($isEntity && !$value instanceof $supportedClass) {
+                $class = $em->getClassMetadata($entityClass);
+                throw new ConstraintDefinitionException(sprintf('The "%s" entity repository does not support the "%s" entity. The entity should be an instance of or extend "%s".', $constraint->entityClass, $class->getName(), $supportedClass));
             }
+            $entityClass = $constraint->entityClass;
+        }
```

Comment:
If we aren't going to try to grab the `$repository` in the block above, we can revert more to the old logic - https://github.com/symfony/symfony/blob/d4b65e44053cbdd44bfdadc08ec7264279c60142/src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php#L125-L138 - i.e. having an `else` to get the repository if `->entityClass` is `null`.

Also, it looks like `$repository` isn't needed until way later. If we could move this block back to where it was before, it would lessen the diff and make this change safer & easier to review.

Reply:
`contains` 

> Checks if the object is part of the current UnitOfWork and therefore managed.

If we check the new Entity, it is not under the control of UnitOfWork. The `contains` method will return false.
Entity would need to be persisted before validation. That's a breaking change and not something that I want.
That's why I'm using  `$em->getRepository`.

I will try to refactor to make the diff as small as possible.

I will change `$isEntity` to `$isValueEntity`.

Reply:
The diff is smaller.